src/sentry/models/outbox.py

```diff
@@ -180,16 +183,39 @@ def next_schedule(self, now: datetime.datetime) -> datetime.datetime:
         return now + (self.last_delay() * 2)
 
     def save(self, **kwds: Any):
+        if _outbox_context.flushing_enabled:
+            transaction.on_commit(lambda: self.drain_shard())
+
         tags = {"category": OutboxCategory(self.category).name}
         metrics.incr("outbox.saved", 1, tags=tags)
         super().save(**kwds)
 
+    @contextlib.contextmanager
+    def process_shard(
+        self, latest_shard_row: OutboxBase | None
+    ) -> Generator[OutboxBase | None, None, None]:
+        flush_all: bool = not bool(latest_shard_row)
+        next_shard_row: OutboxBase | None
+        with transaction.atomic():
```

Comment:
```suggestion
        with transaction.atomic(using=router.db_for_write(self.__class__)):
```
We'll need this to resolve connections in split silo mode as each outbox subclass uses a different connection.

Reply:
Good call!